src/org/zaproxy/zap/extension/pscanrulesAlpha/UsernameIdorScanner.java

```diff
@@ -0,0 +1,205 @@
+		String username;
+
+		for (User user : getUsers()) {
+			username = user.getName();
+			Map<String, String> hashes = new HashMap<String, String>();
+			hashes.put("MD2", DigestUtils.md2Hex(username));
+			hashes.put("MD5", DigestUtils.md5Hex(username));
+			hashes.put("SHA1", DigestUtils.sha1Hex(username));
+			hashes.put("SHA256", DigestUtils.sha256Hex(username));
+			hashes.put("SHA384", DigestUtils.sha384Hex(username));
+			hashes.put("SHA512", DigestUtils.sha512Hex(username));
+			Iterator<Entry<String, String>> iter = hashes.entrySet().iterator();
+			while (iter.hasNext()) {
+				Map.Entry<String, String> pair = iter.next();
+				String hash = pair.getValue();
+				ContentsMatcher hashPatternMatcher = new PatternContentsMatcher(Pattern.compile(hash, Pattern.CASE_INSENSITIVE));
+				String evidence = hashPatternMatcher.match(response);
+				if (evidence != null) {
+					this.raiseAlert(username, evidence, pair.getKey(), id, msg);
+				}
+				iter.remove();
+			}
+		}
+		if (LOGGER.isDebugEnabled()) {
+			LOGGER.debug("\tScan of record " + id + " took " + (System.currentTimeMillis() - start) + " ms");
+		}
+	}
+
+	private void raiseAlert(String username, String evidence, String hashType, int id, HttpMessage msg) {
+		Alert alert = new Alert(getPluginId(), Alert.RISK_INFO, Alert.CONFIDENCE_HIGH, // PluginID, Risk, Reliability
+				getName());
+		alert.setDetail(getDescription(username), // Description
+				msg.getRequestHeader().getURI().toString(), // URI
+				"", // Param
+				"", // Attack
+				getOtherinfo(hashType, evidence), // Other info
+				getSolution(), // Solution
+				getReference(), // References
+				evidence, // Evidence - Return the Server Header info
+				284, // CWE-284: Improper Access Control
+				02, // WASC-02: Insufficient Authorization
+				msg); // HttpMessage
+		parent.raiseAlert(id, alert);
+	}
+
+	@Override
+	public int getPluginId() {
+		return PLUGIN_ID;
+	}
+
+	@Override
+	public String getName() {
+		return Constant.messages.getString(MESSAGE_PREFIX + "name");
+	}
+
+	private String getDescription(String username) {
+		return Constant.messages.getString(MESSAGE_PREFIX + "desc", username);
+	}
+
+	private String getSolution() {
+		return Constant.messages.getString(MESSAGE_PREFIX + "soln");
+	}
+
+	private String getReference() {
+		return Constant.messages.getString(MESSAGE_PREFIX + "refs");
+	}
+
+	private String getOtherinfo(String hashType, String hashValue) {
+		return Constant.messages.getString(MESSAGE_PREFIX + "otherinfo", hashType, hashValue);
+	}
+	
+   private static interface ContentsMatcher {
+        String match(String contents);
+    }
+
+    private static class PatternContentsMatcher implements ContentsMatcher {
+
+        private final Pattern pattern;
+
+        public PatternContentsMatcher(Pattern pattern) {
+            this.pattern = pattern;
+        }
+
+        @Override
+        public String match(String contents) {

```

Comment:
Use of `Pattern` is not bad (I think), it's just plain chars and case insensitive check. The suggestion was just to reduce the current code (use pattern "directly", instead of using the interface and implementation class).